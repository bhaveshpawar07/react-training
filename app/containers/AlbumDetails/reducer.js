/*
 *
 * AlbumDetails reducer
 *
 */
import produce from 'immer';
import { get } from 'lodash';
import { createActions } from 'reduxsauce';

export const initialState = {
  songId: '',
  songData: {},
  songError: null,
  loading: false
};

export const { Types: albumDetailsTypes, Creators: albumDetailsCreators } = createActions({
  requestGetSongDetails: ['songId'],
  successGetSongDetails: ['songData'],
  failureGetSongDetails: ['songError'],
  clearSongDetails: {}
});

export const albumDetailsReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case albumDetailsTypes.REQUEST_GET_SONG_DETAILS:
        draft.songId = action.songId;
        draft.songError = null;
        draft.loading = true;

        break;
      case albumDetailsTypes.SUCCESS_GET_SONG_DETAILS:
        draft.songData = action.songData;
        draft.loading = false;
        break;
      case albumDetailsTypes.FAILURE_GET_SONG_DETAILS:
        draft.songError = get(action.songError, 'message', 'something_went_wrong');
        draft.loading = false;
        break;
      case albumDetailsTypes.CLEAR_SONG_DETAILS:
        draft.songData = {};
        draft.songError = null;
        draft.loading = false;
        break;
      default:
    }
  });

export default albumDetailsReducer;
