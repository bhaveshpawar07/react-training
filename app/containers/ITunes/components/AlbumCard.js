import React from 'react';
import PropTypes from 'prop-types';
import T from '@components/T';
import If from '@components/If';
import { Col } from 'antd';
import styled from 'styled-components';
const CustomCard = styled(Col)`
  && {
    width: 150px;
    display: flex;
    flex-direction: column;
  }
`;
const CustomImg = styled.img`
  && {
    border-radius: ${(props) => props.borderRadius}px;
  }
`;
const CustomDiv = styled.div`
  && {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 10px;
  }
`;
const CustomT = styled(T)`
  && {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    font-weight: ${(props) => props.fontWeight};
    font-size: ${(props) => props.fontSize}px;
  }
`;
export const AlbumCard = ({ artworkUrl100, trackName, trackExplicitness, artistName }) => {
  return (
    <CustomCard span={6}>
      <CustomImg src={artworkUrl100} borderRadius={20} />
      <CustomDiv>
        <CustomT text={trackName} fontWeight={'bold'} />
        <If condition={trackExplicitness === 'explicit'}>
          <CustomImg
            src="https://cdn-icons.flaticon.com/png/512/3097/premium/3097003.png?token=exp=1661161542~hmac=99d22989a8dc8287886e84079bddb458"
            width="15px"
          />
        </If>
      </CustomDiv>
      <CustomT text={artistName} fontSize={11} />
    </CustomCard>
  );
};

AlbumCard.propTypes = {
  artworkUrl100: PropTypes.string,
  trackName: PropTypes.string,
  trackExplicitness: PropTypes.string,
  artistName: PropTypes.string
};

export default AlbumCard;
