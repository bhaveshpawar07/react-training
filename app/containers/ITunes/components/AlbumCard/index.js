import React from 'react';
import PropTypes from 'prop-types';
import T from '@components/T';
import If from '@components/If';
import { Col } from 'antd';
import styled from 'styled-components';
const CustomCard = styled(Col)`
  && {
    width: 150px;
    display: flex;
    flex-direction: column;
  }
`;
const CustomImg = styled.img`
  && {
    border-radius: ${(props) => props.borderRadius}px;
  }
`;
const CustomDiv = styled.div`
  && {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 10px;
  }
`;
const CustomT = styled(T)`
  && {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    font-weight: ${(props) => props.fontWeight};
    font-size: ${(props) => props.fontSize}px;
  }
`;
export const AlbumCard = ({ artworkUrl100, trackName, trackExplicitness, artistName, index, musicPlayer }) => {
  return (
    <CustomCard span={6} data-testid="album-card">
      <CustomImg src={artworkUrl100} borderRadius={20} onClick={() => musicPlayer(index)} />
      <CustomDiv>
        <CustomT data-testid="track-name" text={trackName} fontWeight={'bold'} />
        <If condition={trackExplicitness === 'explicit'}>
          <CustomImg
            src="https://cdn-icons.flaticon.com/png/512/3097/premium/3097003.png?token=exp=1661161542~hmac=99d22989a8dc8287886e84079bddb458"
            width="15px"
          />
        </If>
      </CustomDiv>
      <CustomT data-testid="artist-name" text={artistName} fontSize={11} />
    </CustomCard>
  );
};

AlbumCard.propTypes = {
  artworkUrl100: PropTypes.string,
  trackName: PropTypes.string,
  trackExplicitness: PropTypes.string,
  artistName: PropTypes.string,
  index: PropTypes.number,
  musicPlayer: PropTypes.func
};

export default AlbumCard;
