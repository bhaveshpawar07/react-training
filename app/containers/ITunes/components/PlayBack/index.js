import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { PlayCircleFilled, PauseCircleFilled } from '@ant-design/icons';
import T from '@components/T';
import If from '@components/If';
import { Progress } from 'antd';
import styled from 'styled-components';

const PlayDiv = styled.div`
  && {
    background: black;
    color: white;
    display: flex;
    align-items: center;
    flex-direction: row;
    justify-content: space-between;
    position: fixed;
    bottom: 0;
    width: 690px;
    border-radius: 10px 10px 0px 0px;
    box-shadow: 0px 0px 10px 8px #888888;
  }
`;
const PlayAlbumInfo = styled.div`
  && {
    display: flex;
    align-items: center;
    margin-left: 15px;
  }
`;
const PlayAlbumControl = styled.div`
  && {
    width: 35%;
    margin-right: 35px;
  }
`;

const PlayBack = ({ currentSong, setCurrentSong }) => {
  let audioRef = useRef('');
  return (
    <PlayDiv>
      <PlayAlbumInfo>
        <img src={currentSong.img} />
        <T text={currentSong.songName} style={{ marginLeft: '10px' }} />
      </PlayAlbumInfo>
      <If
        condition={currentSong.play}
        otherwise={
          <PlayCircleFilled
            style={{ fontSize: '25px' }}
            onClick={() => {
              audioRef.current.play();
              setCurrentSong({ ...currentSong, play: true });
            }}
          />
        }
      >
        <PauseCircleFilled
          style={{ fontSize: '25px' }}
          onClick={() => {
            audioRef.current.pause();
            setCurrentSong({ ...currentSong, play: false });
          }}
        />
      </If>
      <PlayAlbumControl>
        <Progress percent={currentSong.progress} status="active" size="small" showInfo={false} />
      </PlayAlbumControl>
      <audio
        autoPlay
        src={currentSong.previewUrl}
        ref={audioRef}
        onEnded={() => setCurrentSong({ ...currentSong, play: false, progress: 0 })}
        onTimeUpdate={() =>
          setCurrentSong({
            ...currentSong,
            progress: (audioRef.current.currentTime / audioRef.current.duration) * 100
          })
        }
      />
    </PlayDiv>
  );
};

PlayBack.propTypes = {
  currentSong: PropTypes.object,
  setCurrentSong: PropTypes.func
};

export default PlayBack;
