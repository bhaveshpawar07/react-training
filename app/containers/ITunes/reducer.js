/*
 *
 * ITunes reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = { tuneName: null, tuneData: {}, tuneError: null };

export const { Types: iTunesTypes, Creators: iTunesCreators } = createActions({
  getItunesData: ['tuneName']
});

export const iTunesReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case iTunesTypes.REQUEST_GET_ITUNES_DATA:
        // eslint-disable-next-line no-console
        console.log('Dispatcher called for ', action.tuneName);
        draft.tuneName = action.tuneName;
        break;
      default:
    }
  });

export default iTunesReducer;
