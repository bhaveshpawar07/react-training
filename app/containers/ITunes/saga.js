import { takeLatest } from 'redux-saga/effects';
import { iTunesTypes } from './reducer';

// Individual exports for testing
const { DEFAULT_ACTION } = iTunesTypes;

export function* defaultFunction(/* action */) {
  // console.log('Do something here')
}

export default function* iTunesSaga() {
  yield takeLatest(DEFAULT_ACTION, defaultFunction);
}
