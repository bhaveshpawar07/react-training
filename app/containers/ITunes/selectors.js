import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the iTunes state domain
 */

const selectITunesDomain = (state) => state.iTunes || initialState;

export const selectITunes = () => createSelector(selectITunesDomain, (substate) => substate);

export const selectSomePayLoad = () => createSelector(selectITunesDomain, (substate) => substate.somePayLoad);
