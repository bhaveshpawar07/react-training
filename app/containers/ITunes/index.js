/**
 *
 * ITunes Container
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, FormattedMessage as T } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { injectSaga } from 'redux-injectors';
import { Card, Input } from 'antd';
import { selectITunesData, selectITunesError, selectITunesName } from './selectors';
import saga from './saga';
import styled from 'styled-components';
import { iTunesCreators } from './reducer';

const { Search } = Input;
const CustomCard = styled(Card)`
  && {
    margin: 20px 0;
    max-width: ${(props) => props.maxwidth};
    color: ${(props) => props.color};
    ${(props) => props.color && `color: ${props.color}`};
  }
`;

const Container = styled.div`
  && {
    display: flex;
    flex-direction: column;
    max-width: ${(props) => props.maxwidth}px;
    width: 100%;
    margin: 0 auto;
    padding: ${(props) => props.padding}px;
  }
`;

export function ITunes({ dispatchGetItunesData }) {
  const handleOnChange = (name) => {
    dispatchGetItunesData(name);
  };

  return (
    <Container maxwidth="500" padding="10">
      <CustomCard title="Album Search">
        <T marginBottom={20} id="Search for any album" />
        <Search
          type="text"
          onChange={(e) => handleOnChange(e.target.value)}
          onSearch={(searchText) => handleOnChange(searchText)}
        />
      </CustomCard>
    </Container>
  );
}

ITunes.propTypes = {
  dispatchGetItunesData: PropTypes.func
};

const mapStateToProps = createStructuredSelector({
  iTunesData: selectITunesData(),
  iTunesName: selectITunesName(),
  iTunesError: selectITunesError()
});

function mapDispatchToProps(dispatch) {
  const { requestGetItunesData, clearItunesData } = iTunesCreators;
  return {
    dispatchGetItunesData: (name) => dispatch(requestGetItunesData(name)),
    dispatchClearItunesData: () => dispatch(clearItunesData())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, injectSaga({ key: 'iTunes', saga }))(ITunes);

export const ITunesTest = compose(injectIntl)(ITunes);
