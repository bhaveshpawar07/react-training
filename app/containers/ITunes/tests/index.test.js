/**
 *
 * Tests for ITunes container
 *
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderProvider, timeout } from '@utils/testUtils';
import { ITunesTest as ITunes } from '../index';
import { iTunesTypes } from '../reducer';
import { mapDispatchToProps } from '@app/containers/ITunes/index';
import { translate } from '@app/components/IntlGlobalProvider';

describe('<ITunes /> container tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ITunes dispatchGetItunesData={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearGithubRepos on empty change and after enter', async () => {
    const getItunesDataSpy = jest.fn();
    const clearItunesDataSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <ITunes dispatchClearItunesData={clearItunesDataSpy} dispatchGetItunesData={getItunesDataSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'test' }
    });
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearItunesDataSpy).toBeCalled();
    fireEvent.keyDown(getByTestId('search-bar'), {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(clearItunesDataSpy).toBeCalledWith();
  });
  it('should call dispatchGetItunesData on some data change and after on enter ', async () => {
    const tuneName = 'test';
    const getItunesDataSpy = jest.fn();
    const clearItunesDataSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <ITunes dispatchClearItunesData={clearItunesDataSpy} dispatchGetItunesData={getItunesDataSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: tuneName }
    });
    await timeout(500);
    expect(getItunesDataSpy).toBeCalledWith(tuneName);
    fireEvent.keyDown(getByTestId('search-bar'), {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
  });
  it('should call dispatchGetItunesData on some data after submit clicked ', async () => {
    const tuneName = 'test';
    const getItunesDataSpy = jest.fn();
    const clearItunesDataSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <ITunes dispatchClearItunesData={clearItunesDataSpy} dispatchGetItunesData={getItunesDataSpy} />
    );
    fireEvent.keyDown(getByTestId('search-bar'), { keyCode: 13, target: { value: tuneName } });
    await timeout(500);
    expect(getItunesDataSpy).toBeCalledWith(tuneName);
  });
  it('should  dispatchGetItunesData on update on mount if iTunesName is already persisted', async () => {
    const iTunesName = 'test';
    renderProvider(<ITunes iTunesName={iTunesName} iTunesData={null} dispatchGetItunesData={submitSpy} />);

    await timeout(500);
    expect(submitSpy).toBeCalledWith(iTunesName);
  });
  it('should validate mapDispatchToProps actions', async () => {
    const dispatchItunesSearchSpy = jest.fn();
    const tuneName = 'test';
    const actions = {
      dispatchGetItunesData: { tuneName, type: iTunesTypes.REQUEST_GET_ITUNES_DATA },
      dispatchClearItunesData: { type: iTunesTypes.CLEAR_ITUNES_DATA }
    };

    const props = mapDispatchToProps(dispatchItunesSearchSpy);
    props.dispatchGetItunesData(tuneName);
    expect(dispatchItunesSearchSpy).toHaveBeenCalledWith(actions.dispatchGetItunesData);

    await timeout(500);
    props.dispatchClearItunesData();
    expect(dispatchItunesSearchSpy).toHaveBeenCalledWith(actions.dispatchClearItunesData);
  });
  it('should render data, when data is present', async () => {
    const iTunesData = { results: [{ name: 'test' }] };
    const { getByTestId } = renderProvider(<ITunes iTunesData={iTunesData} dispatchGetItunesData={submitSpy} />);
    expect(getByTestId('dataRow')).toBeInTheDocument();
  });
  it('should render Skeleton when loading is true', async () => {
    const iTunesName = 'test';
    const { getByTestId, baseElement } = renderProvider(<ITunes dispatchGetItunesData={submitSpy} loading={true} />);
    fireEvent.change(getByTestId('search-bar'), { target: { value: iTunesName } });
    await timeout(500);
    expect(baseElement.getElementsByClassName('ant-skeleton').length).toBe(1);
  });
  it('should render error message, when api returns some error', () => {
    const iTunesError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(<ITunes iTunesError={iTunesError} />);
    // expect(getByTestId('error-message').textContent).toBe(iTunesError);
    expect(getByTestId('error-message')).toHaveTextContent(iTunesError);
  });
  it('should show music playback after a album is clicked for the same album', async () => {
    const trackName = 'test';
    const iTunesData = { results: [{ trackName: trackName }] };
    const { getByTestId } = renderProvider(<ITunes iTunesData={iTunesData} dispatchGetItunesData={submitSpy} />);
    expect(getByTestId('dataRow')).toBeInTheDocument();
    expect(getByTestId('album-card')).toBeInTheDocument();
    await timeout(500);
    fireEvent.click(getByTestId('album-card'));
    expect(getByTestId('playback-control')).toBeInTheDocument();
    expect(getByTestId('playback-songName')).toHaveTextContent(trackName);
  });

  it('should show play button when pause is clicked and pause button when play is clicked', async () => {
    const trackName = 'test';
    const iTunesData = { results: [{ trackName: trackName }] };
    const { getByTestId } = renderProvider(<ITunes iTunesData={iTunesData} dispatchGetItunesData={submitSpy} />);

    expect(getByTestId('dataRow')).toBeInTheDocument();
    expect(getByTestId('album-card')).toBeInTheDocument();

    await timeout(500);
    fireEvent.click(getByTestId('album-card'));

    expect(getByTestId('playback-control')).toBeInTheDocument();
    expect(getByTestId('pause')).toBeInTheDocument();

    fireEvent.click(getByTestId('pause'));

    expect(getByTestId('play')).toBeInTheDocument();

    fireEvent.click(getByTestId('play'));

    expect(getByTestId('pause')).toBeInTheDocument();

    expect(getByTestId('pause')).toBeInTheDocument();
    const audioElement = getByTestId('audioElement');
    audioElement.dispatchEvent(new Event('ended'));
    expect(getByTestId('play')).toBeInTheDocument();
  });
});
