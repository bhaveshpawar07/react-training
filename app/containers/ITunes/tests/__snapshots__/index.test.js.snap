// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunes /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      ITunes
    </div>
  </div>
</body>
`;
